File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-1.py
from collections import defaultdict
class Solution:
    # top-down DP solution, O(n) time, O(n) space
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        cache = defaultdict(int)

        def dfs(i):
            if i >= len(cost):
                return 0
            
            if i in cache:
                return cache[i]

            curr_cost = cost[i] + min(dfs(i + 1), dfs(i + 2))
            cache[i] = curr_cost
            return curr_cost

        return min(dfs(0), dfs(1))

File: Data Structures & Algorithms/pascals-triangle-ii/submission-1.cpp
class Solution {
public:
    // top-down DP, O(n^2) time, O(n) space
    vector<int> getRow(int rowIndex) {
        if (rowIndex == 0) {
            return {1};
        }

        vector<int> currRow = {1};
        vector<int> prevRow = getRow(rowIndex - 1);

        for (int i = 1; i < rowIndex; i++) {
            currRow.push_back(prevRow[i] + prevRow[i - 1]);
        }

        currRow.push_back(1);
        return currRow;
    }
};